### DISCUSSION TITLE
properly drawing ISO/BAE enter key?

### DISCUSSION THREAD

**Discussion by zackforbing** [nickcoutsos/keymap-editor#196] [Q&A]
I found the keymap layout helper app (which could be more easily discoverable!), but I'm attempting to model an ISO enter, and it doesn't seem that there is a way to show the "step", it only reads x/h/w and not x2/h2/w2, which is how the secondary "step" is denoted. is there a way I'm missing to model keys like this in <board>.json? 

I currently just model two separate keys occupying the same row/col with the larger one on top:
![image](https://github.com/nickcoutsos/keymap-editor/assets/13476440/9843ed4a-9842-4123-9013-548814c3e946)

```
...
{ "row": 3, "col": 8, "x":   8.5, "y":    3, "w":  1.5, "h": 1 },
{ "row": 3, "col": 8, "x":  8.75, "y":    3, "w": 1.25, "h": 2, "2h": 1 }
...
```

but this throws an error on opening the layout in Keymap Editor.

**Comment by nickcoutsos** [comment]
It doesn't exist currently. I mocked something up a while ago but it relies on CSS clipping which has other effects that made me unhappy.

> x2/h2/w2, which is how the secondary "step" is denoted

I haven't heard of this, it's not something I support and it doesn't seem to be specified in QMK's layout format either.

**Reply by zackforbing** [reply]
it wasn't QMK's format, but keyboard-layout-editor's format. QMK wouldn't really have a need for those designations, IMO, since it isn't making a graphical representation of the keys at any point, really.

looking at the [default ISO 60% kle](http://www.keyboard-layout-editor.com/##@@=%C2%AC%0A%60&=!%0A1&=%22%0A2&=%C2%A3%0A3&=$%0A4&=%25%0A5&=%5E%0A6&=%2F&%0A7&=*%0A8&=(%0A9&=)%0A0&=%2F_%0A-&=+%0A%2F=&_w:2%3B&=Backspace%3B&@_w:1.5%3B&=Tab&=Q&=W&=E&=R&=T&=Y&=U&=I&=O&=P&=%7B%0A%5B&=%7D%0A%5D&_x:0.25&w:1.25&h:2&w2:1.5&h2:1&x2:-0.25%3B&=Enter%3B&@_w:1.75%3B&=Caps%20Lock&=A&=S&=D&=F&=G&=H&=J&=K&=L&=%2F:%0A%2F%3B&=%2F@%0A'&=~%0A%23%3B&@_w:1.25%3B&=Shift&=%7C%0A%5C&=Z&=X&=C&=V&=B&=N&=M&=%3C%0A,&=%3E%0A.&=%3F%0A%2F%2F&_w:2.75%3B&=Shift%3B&@_w:1.25%3B&=Ctrl&_w:1.25%3B&=Win&_w:1.25%3B&=Alt&_a:7&w:6.25%3B&=&_a:4&w:1.25%3B&=AltGr&_w:1.25%3B&=Win&_w:1.25%3B&=Menu&_w:1.25%3B&=Ctrl), if you click the raw data tab, you can see those fields in use for the enter key.

**Reply by nickcoutsos** [reply]
Ah, ok. QMK based the [layout format](https://docs.qmk.fm/#/reference_info_json?id=layouts) for QMK configurator on KLE but more clearly defined, which is why I base my format on theirs.

**Comment by nickcoutsos** [comment]
I pulled up my old branch to play around with:

![Closeup of a keyboard layout centered on a ISO-shaped Enter key](https://github.com/nickcoutsos/keymap-editor/assets/1165066/a991e364-611a-4514-82c3-04430e9b3ffd)

There are some issues as I mentioned earlier due to the ISO Enter shape being a purely cosmetic styling, while the geometry is still treated as a 1.5x2 rectangle.

![Closeup of a keyboard layout centered on a ISO-shaped Enter key with a rectangular red dotted border indicating an overlap with a neighboring key](https://github.com/nickcoutsos/keymap-editor/assets/1165066/f3158200-d308-4e7a-9f56-e5d8bb7c9d2e)

![Closeup of a keyboard layout centered on a ISO-shaped Enter key with a rectangular red dotted border indicating an overlap with a neighboring key and a blue border indicating overlap with a drag-selection operation](https://github.com/nickcoutsos/keymap-editor/assets/1165066/d02aaa29-33d2-4abd-8d8d-65a5cbb98576)

This means that the overlap check (a warning specific to the helper app) and drag selections are going to trigger in situations where you're in that region but not over the actual key polygon. That's a fairly small space so it shouldn't interfere too badly with regular use.

It also means that any border styling I want to apply is cut off. In the helper app that's a small annoyance since it's a visual cue for pending key selection. In keymap editor this isn't a factor because none of the key styles (as far as I recall) use borders, but if I were to go forward with this it would likely nag at me.
